src/openMVG/localization/VoctreeLocalizer.cpp

```diff
@@ -1038,13 +1040,410 @@ bool VoctreeLocalizer::robustMatching(matching::RegionsMatcherT<MatcherT> & matc
+{
+  const size_t numCams = vec_imageGrey.size();
+  assert(numCams == vec_queryIntrinsics.size());
+  assert(numCams == vec_subPoses.size() + 1);
+
+  std::vector<std::unique_ptr<features::Regions> > vec_queryRegions(numCams);
+  std::vector<std::pair<std::size_t, std::size_t> > vec_imageSize;
+  
+  //@todo parallelize?
+  for(size_t i = 0; i < numCams; ++i)
+  {
+    // extract descriptors and features from each image
+    vec_queryRegions[i] = std::unique_ptr<features::Regions>(new features::SIFT_Regions());
+    POPART_COUT("[features]\tExtract SIFT from query image...");
+    _image_describer->Describe(vec_imageGrey[i], vec_queryRegions[i]);
+    POPART_COUT("[features]\tExtract SIFT done: found " <<  vec_queryRegions[i]->RegionCount() << " features");
+    // add the image size for this image
+    vec_imageSize.emplace_back(vec_imageGrey[i].Width(), vec_imageGrey[i].Height());
+}
+  assert(vec_imageSize.size() == vec_queryRegions.size());
+          
+  return localizeRig(vec_queryRegions,
+                     vec_imageSize,
+                     parameters,
+                     vec_queryIntrinsics,
+                     vec_subPoses,
+                     rigPose,
+                     vec_locResults);
+}
+
+
+bool VoctreeLocalizer::localizeRig(const std::vector<std::unique_ptr<features::Regions> > & vec_queryRegions,
+                                   const std::vector<std::pair<std::size_t, std::size_t> > &vec_imageSize,
+                                   const LocalizerParameters *parameters,
+                                   std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
+                                   const std::vector<geometry::Pose3 > &vec_subPoses,
+                                   geometry::Pose3 &rigPose,
+                                   std::vector<LocalizationResult>& vec_locResults)
+{
+#ifdef HAVE_OPENGV
+  return localizeRig_opengv(vec_queryRegions,
+                            vec_imageSize,
+                            parameters,
+                            vec_queryIntrinsics,
+                            vec_subPoses,
+                            rigPose,
+                            vec_locResults);
+#else
+  return localizeRig_naive(vec_queryRegions,
+                           vec_imageSize,
+                           parameters,
+                           vec_queryIntrinsics,
+                           vec_subPoses,
+                           rigPose,
+                           vec_locResults);
+#endif
+}
+
+
+#ifdef HAVE_OPENGV
+
+bool VoctreeLocalizer::localizeRig_opengv(const std::vector<std::unique_ptr<features::Regions> > & vec_queryRegions,
+                                          const std::vector<std::pair<std::size_t, std::size_t> > &vec_imageSize,
+                                          const LocalizerParameters *parameters,
+                                          std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
```

Comment:
Use boost::ptr_vector<IntrinsicBase> instead of std::vector<cameras::Pinhole_Intrinsic_Radial_K3 >.